core/embed/rust/src/trezorhal/storage.rs

```diff
@@ -0,0 +1,330 @@
+use super::ffi;
+use crate::error::Error;
+use core::ptr;
+use cstr_core::{cstr, CStr};
+
+/// Result of PIN delay callback.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum PinCallbackResult {
+    /// Continue waiting for PIN unlock.
+    Continue,
+    /// Abort the unlock attempt before time is up.
+    Abort,
+}
+
+/// PIN delay callback function type.
+/// The storage layer will call this function while the PIN timeout is in
+/// progress. This is useful for showing UI progress bar.
+/// `wait` is the total number of seconds waiting.
+/// `progress` is a value between 0 and 1000, where 1000 indicates 100%.
+/// `message` is a message to show to the user.
+pub type PinDelayCallback = fn(wait: u32, progress: u32, message: &str) -> PinCallbackResult;
+
+pub type ExternalSalt = [u8; ffi::EXTERNAL_SALT_SIZE as usize];
+
+/// Static reference to the currently set PIN callback function.
+static mut PIN_UI_CALLBACK: Option<PinDelayCallback> = None;
+
+/// C-compatible wrapper for the Rust callback.
+unsafe extern "C" fn callback_wrapper(
+    wait: u32,
+    progress: u32,
+    message: *const cty::c_char,
+) -> ffi::secbool {
+    let message = unsafe { CStr::from_ptr(message as _) };
+    let result = unsafe {
+        PIN_UI_CALLBACK
+            .map(|c| c(wait, progress, message.to_str().unwrap_or("")))
+            .unwrap_or(PinCallbackResult::Continue)
+    };
+    if matches!(result, PinCallbackResult::Abort) {
+        ffi::sectrue
+    } else {
+        ffi::secfalse
+    }
+}
+
+/// Storage error type.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum StorageError {
+    /// Data is too long or not appropriate for the operation.
+    InvalidData,
+    /// Write failed.
+    WriteFailed,
+    /// Read failed.
+    ReadFailed,
+    /// Delete failed.
+    DeleteFailed,
+    /// Failed to perform the get-and-increment operation on a counter.
+    CounterFailed,
```

Comment:
Note: it kind of looks like that read/write/delete/counter could be folded into a single `OperationFailed`, because the only difference is error string. (`InvalidData` is different in that it is a problem with the data passed in, not with the storage operation)

but I am not sure how these are used in `storage_field` so let's keep them for now, or else we'll remove them when merging `storage_field`